ports/unix/main.c

```diff
@@ -297,22 +226,43 @@ static int do_repl(void) {
             line = line3;
         }
 
-        int ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
+        ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
         free(line);
         if (ret & FORCED_EXIT) {
             return ret;
         }
     }
-
     #endif
+    return ret;
 }
 
+
 static int do_file(const char *file) {
-    return execute_from_lexer(LEX_SRC_FILENAME, file, MP_PARSE_FILE_INPUT, false);
+    int ret = pyexec_file(file);
+    // pyexec returns 1 for success, 0 for exception, PYEXEC_FORCED_EXIT for SystemExit
+    // Convert to unix port's expected codes: 0 for success, 1 for exception, FORCED_EXIT|val for SystemExit
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret & PYEXEC_FORCED_EXIT) {
+        return ret; // SystemExit with exit value in lower 8 bits
+    } else {
+        return 1; // exception
+    }
 }
 
 static int do_str(const char *str) {
-    return execute_from_lexer(LEX_SRC_STR, str, MP_PARSE_FILE_INPUT, false);
+    vstr_t vstr;
+    vstr_init(&vstr, strlen(str));
+    vstr_add_strn(&vstr, str, strlen(str));
+    int ret = pyexec_vstr(&vstr, false);
+    vstr_clear(&vstr);
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret & PYEXEC_FORCED_EXIT) {
+        return ret; // SystemExit with exit value in lower 8 bits
+    } else {
+        return 1; // exception
+    }
```

Comment:
This code should be factored/shared with `do_file()`.

Reply:
Updated as suggested thanks

Reply:
I think you forgot to commit this change?